qiskit/quantum_info/synthesis/two_qubit_kak.py

```diff
@@ -0,0 +1,350 @@
+    # U[1, 0] = exp(i(phi-lambda)/2) * sin(theta/2)
+    # U[1, 1] = exp(i(phi+lambda)/2) * cos(theta/2)
+    # Find theta
+    if abs(U[0, 0]) > _CUTOFF_PRECISION:
+        theta = 2 * math.acos(abs(U[0, 0]))
+    else:
+        theta = 2 * math.asin(abs(U[1, 0]))
+    # Find phi and lambda
+    phase11 = 0.0
+    phase10 = 0.0
+    if abs(math.cos(theta/2.0)) > _CUTOFF_PRECISION:
+        phase11 = U[1, 1] / math.cos(theta/2.0)
+    if abs(math.sin(theta/2.0)) > _CUTOFF_PRECISION:
+        phase10 = U[1, 0] / math.sin(theta/2.0)
+    phiplambda = 2 * math.atan2(np.imag(phase11), np.real(phase11))
+    phimlambda = 2 * math.atan2(np.imag(phase10), np.real(phase10))
+    phi = 0.0
+    if abs(U[0, 0]) > _CUTOFF_PRECISION and abs(U[1, 0]) > _CUTOFF_PRECISION:
+        phi = (phiplambda + phimlambda) / 2.0
+        lamb = (phiplambda - phimlambda) / 2.0
+    else:
+        if abs(U[0, 0]) < _CUTOFF_PRECISION:
+            lamb = -phimlambda
+        else:
+            lamb = phiplambda
+    # Check the solution
+    Rzphi = np.array([[np.exp(-1j*phi/2.0), 0],
+                      [0, np.exp(1j*phi/2.0)]], dtype=complex)
+    Rytheta = np.array([[np.cos(theta/2.0), -np.sin(theta/2.0)],
+                        [np.sin(theta/2.0), np.cos(theta/2.0)]], dtype=complex)
+    Rzlambda = np.array([[np.exp(-1j*lamb/2.0), 0],
+                         [0, np.exp(1j*lamb/2.0)]], dtype=complex)
+    V = np.dot(Rzphi, np.dot(Rytheta, Rzlambda))
+    if la.norm(V - U) > _CUTOFF_PRECISION:
+        raise QiskitError("euler_angles_1q: incorrect result")
+    return theta, phi, lamb
+
+
+def simplify_U(theta, phi, lam):
+    """Return the gate u1, u2, or u3 implementing U with the fewest pulses.
+
+    The returned gate implements U exactly, not up to a global phase.
+
+    Args:
+        theta, phi, lam: input Euler rotation angles for a general U gate
+
+    Returns:
+        Gate: one of IdGate, U1Gate, U2Gate, U3Gate.
+    """
+    gate = U3Gate(theta, phi, lam)
+    # Y rotation is 0 mod 2*pi, so the gate is a u1
+    if abs(gate.params[0] % (2.0 * math.pi)) < _CUTOFF_PRECISION:
+        gate = U1Gate(gate.params[0] + gate.params[1] + gate.params[2])
+    # Y rotation is pi/2 or -pi/2 mod 2*pi, so the gate is a u2
+    if isinstance(gate, U3Gate):
+        # theta = pi/2 + 2*k*pi
+        if abs((gate.params[0] - math.pi / 2) % (2.0 * math.pi)) < _CUTOFF_PRECISION:
+            gate = U2Gate(gate.params[1],
+                          gate.params[2] + (gate.params[0] - math.pi / 2))
+        # theta = -pi/2 + 2*k*pi
+        if abs((gate.params[0] + math.pi / 2) % (2.0 * math.pi)) < _CUTOFF_PRECISION:
+            gate = U2Gate(gate.params[1] + math.pi,
+                          gate.params[2] - math.pi + (gate.params[0] + math.pi / 2))
+    # u1 and lambda is 0 mod 4*pi so gate is nop
+    if isinstance(gate, U1Gate) and abs(gate.params[0] % (4.0 * math.pi)) < _CUTOFF_PRECISION:
+        gate = IdGate()
+    return gate
+
+
+def two_qubit_kak(unitary):
```

Comment:
This function should be able to take a list, numpy array, Operator, Channel, Unitary etc as input. As it is it will raise an error for anything but a Unitary object.

You can do something like:

```python
if not isinstance(unitary, (Unitary, Operator))
    unitary = Operator(unitary)
    if not unitary.is_unitary():
        raise QiskitError(input is not unitary).
mat = unitary.data  # for operator (Unitary class still needs to be changed to have same interface as operator)

Reply:
are you sure? why not just clearly define the interfaces here (i.e. kak decomposition takes in a unitary, which in Qiskit is class Unitary). 

My immediate problem is that this causes circular import hell (Unitary needs to KAK to define a decomposition, and KAK needs Unitary to check input instance type). But if you really want it I might be able to make it work somehow.

Reply:
Once unitary inherits from Operator you won't have this problem. You will just do `issubclass(obj.__class__, BaseOperator)` or something like that.